pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,107 @@
+                               BiFunction<
+                                       DefaultBlockParameter,
+                                       DefaultBlockParameter,
+                                       List<DepositContract.DepositEventEventResponse>> depositEventsInRangeFunction) {
+    this.web3j = web3j;
+    this.depositsFromBlockEventConsumer = depositsFromBlockEventConsumer;
+    this.depositEventsInRangeFunction = depositEventsInRangeFunction;
+  }
+
+  public void start() {
+    newBlockSubscription = web3j
+            .blockFlowable(false)
+            .map(EthBlock::getBlock)
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(this::getLatestDeposits);
+  }
+
+  public void stop() {
+    newBlockSubscription.dispose();
+  }
+
+  private void getLatestDeposits(BigInteger latestCanonicalBlockNumber) {
+    DefaultBlockParameter toBlock = DefaultBlockParameter.valueOf(latestCanonicalBlockNumber);
+    DefaultBlockParameter fromBlock;
+    if (latestSuccessfullyQueriedBlockNumber.equals(BigInteger.ZERO)) {
+      fromBlock = DefaultBlockParameterName.EARLIEST;
+    } else {
+      fromBlock = DefaultBlockParameter.valueOf(latestSuccessfullyQueriedBlockNumber);
+    }
+
+    List<DepositContract.DepositEventEventResponse> events = depositEventsInRangeFunction.apply(fromBlock, toBlock);
+
+    List<SafeFuture<DepositsFromBlockEvent>> depositsFromEventFutures = events.stream()
+            .collect(Collectors.groupingBy(event -> event.log.getBlockNumber(), TreeMap::new, Collectors.toList()))
+            .values()
+            .stream()
+            .map(depositsEventResponseList -> {
+              List<Deposit> depositsList = depositsEventResponseList.stream().map(Deposit::new).collect(Collectors.toList());
+              return SafeFuture.of(web3j.ethGetBlockByHash(
+                      depositsEventResponseList.get(0).log.getBlockHash(),
+                      false
+              )
+                      .sendAsync()
+                      .thenApply(EthBlock::getBlock)
+                      .thenApply(block -> new DepositsFromBlockEvent(
+                              UnsignedLong.valueOf(block.getNumber()),
+                              Bytes32.fromHexString(block.getHash()),
+                              UnsignedLong.valueOf(block.getTimestamp()),
+                              depositsList)));
+            }).collect(Collectors.toList());
+
+    SafeFuture.allOf(depositsFromEventFutures.toArray(SafeFuture[]::new))
+            .thenRun(() -> depositsFromEventFutures
+                    .stream()
+                    .map(future -> checkNotNull(future.getNow(null)))
+                    .sorted(Comparator.comparing(DepositsFromBlockEvent::getBlockNumber))
+                    .forEachOrdered(this::publishDeposits))
+            .thenRun(() -> latestSuccessfullyQueriedBlockNumber = latestCanonicalBlockNumber);
```

Comment:
This probably should be `finish` as `thenRun` returns another `SafeFuture`.  And we probably should handle errors here explicitly - they likely just get logged and we retry but that's better than falling back to the default exception handler.

Reply:
Done.